// magnet/build/build/client.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildClient = buildClient;

var _metal = require('metal');

var _log = require('../log');

var _log2 = _interopRequireDefault(_log);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Builds client using webpack config.
 * @param {Magnet} magnet
 */
async function buildClient(magnet) {
  _log2.default.info(false, 'Building client…');
  await maybeSetupPluginsWebpack_(magnet);
  await maybeSetupMagnetWebpack_(magnet);
  if (Object.keys(magnet.webpackConfig.entry).length) {
    await runWebpack_(magnet);
  }
}

/**
 * Maybe run webpack if it has entries.
 * @param {Magnet} magnet
 * @return {Promise}
 * @private
 */
function runWebpack_(magnet) {
  return new Promise((resolve, reject) => {
    (0, _webpack2.default)(magnet.webpackConfig, (err, stats) => {
      if (err) {
        _log2.default.error(false, err);
        reject(err);
      }
      const output = stats.toString({
        colors: true,
        chunks: false
      });
      resolve(output);
    });
  });
}

/**
 * Maybe setup Plugins webpack configuration.
 * @param {Magnet} magnet
 * @private
 */
async function maybeSetupPluginsWebpack_(magnet) {
  try {
    for (const plugin of magnet.getPlugins()) {
      if ((0, _metal.isFunction)(plugin.webpackConfig)) {
        magnet.webpackConfig = await plugin.webpackConfig(magnet.webpackConfig, magnet);
      }
    }
  } catch (error) {
    _log2.default.error(false, error);
  }
}

/**
 * Maybe applies magnet webpack config.
 * @param {Magnet} magnet
 * @private
 */
async function maybeSetupMagnetWebpack_(magnet) {
  const magnetConfig = magnet.getConfig().magnet;

  if ((0, _metal.isFunction)(magnetConfig.webpack)) {
    magnet.webpackConfig = await magnetConfig.webpack(magnet.webpackConfig, magnet);
  }
}
