// magnet/build/routes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRoutesTable = getRoutesTable;

var _magnet = require('./magnet');

var _magnet2 = _interopRequireDefault(_magnet);

var _cliTable = require('cli-table');

var _cliTable2 = _interopRequireDefault(_cliTable);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _magnetPluginController = require('magnet-plugin-controller');

var _magnetPluginController2 = _interopRequireDefault(_magnetPluginController);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Scans build files collecting routes metadata into an array consumed
 * by `magnet routes` command output table.
 * @param {Magnet} magnet
 * @return {Array.<Array.<string>>}
 * @private
 */
function getRoutesDefinition_(magnet) {
  let dist = magnet.getServerDistDirectory();
  let files = magnet.getFiles({ directory: dist, realpath: true });

  let results = [];
  files.forEach(file => {
    switch (file) {
      case _path2.default.join(dist, _magnet2.default.LifecyleFiles.START):
      case _path2.default.join(dist, _magnet2.default.LifecyleFiles.STOP):
        return;
    }

    let module = require(file);
    let short = file.substring(dist.length);

    if (_magnetPluginController2.default.test(module, file, magnet)) {
      results.push(['―', '―', '―', short]);
    } else {
      let route = module.route;
      if (route) {
        let path = route.path;
        let method = route.method || 'get';
        let type = route.type || 'html';
        results.push([method.toUpperCase(), path, type, short]);
      }
    }
  });
  return results;
}

/**
 * Builds routing table based on route definitions. If no routes were found
 * returns empty string, otherwise return as example below:
 * ┌────────┬──────┬──────┬───────────┐
 * │ method │ path │ type │ file      │
 * ├────────┼──────┼──────┼───────────┤
 * │ POST   │ /api │ json │ /api.js   │
 * │ GET    │ /    │ html │ /index.js │
 * └────────┴──────┴──────┴───────────┘
 * @param {Magnet} magnet
 * @return {String}
 */
function getRoutesTable(magnet) {
  const table = new _cliTable2.default({
    head: ['method', 'path', 'type', 'file'],
    style: {
      border: ['gray'],
      compact: true,
      head: ['gray']
    }
  });
  table.push(...getRoutesDefinition_(magnet));
  if (table.length) {
    return table.toString();
  }
  return '';
}
